### DISCUSSION TITLE
Recent Files fzf Picker

### DISCUSSION THREAD

**Discussion by Randalix** [helix-editor/helix#4802] [Show and tell]
I hacked together a script to open recent files in helix. Maybe it is useful to someone.
Instead of opening helix, you run this script. You can alias it in your .bashrc .
If someone has improvements, feel free to share.
```
#!/usr/bin/env bash
SED="sed"
HX=hx
if [[ "$OSTYPE" == "darwin"* ]]; then
		SED="gsed"
fi
RECENT=~/.cache/recent_hx
FILE=${@:1}
if [ ! "$FILE" ]; then
	FILE=$(tac $RECENT  | fzf)
fi
if [ "$FILE" ]; then
	FILE=$(printf "%s" "$(cd "$(dirname -- "$FILE")" >/dev/null; pwd -P)/$(basename -- "$@")"
 $FILE)
	echo $FILE
	$SED -i "\#$FILE#d" $RECENT
	echo $FILE>>$RECENT
fi
$HX $FILE

```

**Comment by robrecord** [comment]
Thanks for this.
How does this cache get populated? For me, the path doesn't exist.

**Comment by David-Else** [comment]
I am working on something myself: [kitty-terminal-project-launcher](https://github.com/David-Else/kitty-terminal-project-launcher)
## kitty-terminal-project-launcher
## Features:

- Fuzzy search for all or just recent projects using [fzf](https://github.com/junegunn/fzf).
- Open projects with your default terminal editor. < Helix of course :)
- Open preconfigured sets of named windows for different project types.
- Automatic closing of windows when exiting your editor.
- Create new project folders based on your chosen directory structure

It would be easy to strip out the Kitty bits... but much better to just choose Kitty as your terminal... as it is the ultimate terminal on planet earth :)

I think the documentation makes it sound more complicated than it is, @Randalix I did not know about the `tac` command! I may be able to use to simplify some of my logic.

**Comment by webdev23** [comment]
Alternatively, using your shell, we could search for the last file using for example ripgrep:

Using an Alias:

`hx ~/.zshrc`

`alias mod='rg --files --hidden --stats | xargs -d "\n" stat --format="%Y %n" 2>/dev/null | sort -n | tail -n 3 | cut -d" " -f2-'`

`source ~/.zshrc`

Now, from any directory, this open in Helix the 3 last modified file of the whole tree of sub directories:

`hx $(mod)`

Modify the above `tail` command for the amount of files to open.

**Comment by zetashift** [comment]
For the curious. using the Steel build you could use the existing file picker to create a recent file picker like so:
https://github.com/mattwparas/helix-config/blob/master/cogs/recentf.scm

**Reply by markstos** [reply]
Could it be used to implement "z.vim"? It caches frequent/recent directories visited and makes them easy to jump to. The same cache is shared between the shell and editors, which is nice. https://github.com/lingceng/z.vim

**Reply by zetashift** [reply]
That vimscript is undecipherable for me, sorry :sweat: .
But to answer your question, since `z` is doing all the heavy lifting here, I don't see why it wouldn't be possible. However, some funky stuff might happen if you change the current working directory in your current helix instance. So for jumping to a dir I would recommend to open a new split/tab in your window manager or terminal multiplexer.

**Reply by zetashift** [reply]
I believe I pressed enter too quickly. If you use a terminal multiplexer(recommended imho!) you could script this plugin outside of helix, if you go the route of opening the directory in a new tab/pane:
- map a keybind to a script which:
- fuzzy lists(using `fzf`) the recent entries from, `zoxide`/`z` has this builtin if you run the command `zi`.
- with a chosen directory open a new pane/tab and run the command `hx chosen-dir`.

**Reply by markstos** [reply]
Thanks for the tip @zetashift I see a number of people talking about Zellij now. I used to use `tmux` and `screen` before that. Now my tiling window manager servers that role.

Yes, it would be easy  to write a dmenu picker script for `z` to launch a new Helix instance:

>  wezterm -e fish -c  "cd (z --list | perl -pe 's/^[\d\.]+\s+//' | dmenu); hx"

But often I want to `:cd` within the same helix session. The Neovim plugin allowed me to use `:Z` within Neovim the same way I used `z` in the shell. 

In Helix, it might be more natural to remember the recent/frequent list of directories, and have it's own picker when using `cd`. If Helix had the option to share the same cache file and format as Z, that would be 💯
